core/server/influx-metrics.spec.js

```diff
@@ -0,0 +1,204 @@
+
+      expect(statusCode).to.be.equal(200)
+      expect(body).to.contain(
+        'prometheus,application=shields,env=test-env,instance=instance2 counter1=11'
+      )
+      expect(headers).to.have.property('content-type', 'text/plain')
+    })
+  })
+
+  describe('pushing component', function() {
+    let influxMetrics
+    beforeEach(function() {
+      sinon.spy(log, 'error')
+    })
+    afterEach(function() {
+      log.error.restore()
+      influxMetrics.stopPushingMetrics()
+      nock.cleanAll()
+      nock.enableNetConnect()
+    })
+    it('should send metrics', async function() {
+      const scope = nock('http://shields-metrics.io/', {
+        reqheaders: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+      })
+        .persist()
+        .post(
+          '/metrics',
+          'prometheus,application=shields,env=test-env,instance=instance2 counter1=11'
+        )
+        .basicAuth({ user: 'metrics-username', pass: 'metrics-password' })
+        .reply(200)
+      influxMetrics = new InfluxMetrics(metricInstance, instanceMetadata, {
+        uri: 'http://shields-metrics.io/metrics',
+        timeoutMillseconds: 100,
+        intervalSeconds: 0,
+        username: 'metrics-username',
+        password: 'metrics-password',
+      })
+
+      await influxMetrics.startPushingMetrics()
+
+      await waitForExpect(
+        () => {
+          expect(scope.isDone()).to.be.equal(
+            true,
+            `pending mocks: ${scope.pendingMocks()}`
+          )
+        },
+        1000,
+        1
+      )
+    })
+
+    it('should log errors', async function() {
+      nock.disableNetConnect()
+      influxMetrics = new InfluxMetrics(metricInstance, instanceMetadata, {
+        uri: 'http://shields-metrics.io/metrics',
+        timeoutMillseconds: 50,
+        intervalSeconds: 0,
+        username: 'metrics-username',
+        password: 'metrics-password',
+      })
+
+      influxMetrics.startPushingMetrics()
+
+      await waitForExpect(
+        () => {
+          expect(log.error).to.be.calledWith(
+            sinon.match
+              .instanceOf(Error)
+              .and(
+                sinon.match.has(
+                  'message',
+                  'Cannot push metrics. Cause: NetConnectNotAllowedError: Nock: Disallowed net connect for "shields-metrics.io:80/metrics"'
+                )
+              )
+          )
+        },
+        1000,
+        1
+      )
+    })
+
+    it('should log error responses', async function() {
+      nock('http://shields-metrics.io/')
+        .persist()
+        .post('/metrics')
+        .reply(400)
+      influxMetrics = new InfluxMetrics(metricInstance, instanceMetadata, {
+        uri: 'http://shields-metrics.io/metrics',
+        timeoutMillseconds: 50,
+        intervalSeconds: 0,
+        username: 'metrics-username',
+        password: 'metrics-password',
+      })
+
+      influxMetrics.startPushingMetrics()
+
+      await waitForExpect(
```

Comment:
This would be easier to test if you extract `sendMetrics` to a method which can be invoked directly and returns a promise.

Reply:
`sendMetrics` returning a promise is a good idea and test are simpler now. I left one test with `waitForExpect` to test `influxMetrics.startPushingMetrics`.

Reply:
Would you be comfortable using a mock instead?

Reply:
A mock should be OK, but I think `sinon.useFakeTimers` + `tickAsync` is slightly better. What do you think about this approach? Of course I can change it to use mocks.